### PR TITLE
JWKS support

### PR DESCRIPTION
**Pull request by afflerbach**

### CODE REVIEW

File: accesscontrol/jwt.go [hunk 1]
```diff
@@ -167,7 +196,24 @@ func (j *JWT) Validate(req *http.Request) error {
 	return nil
 }
 
-func (j *JWT) getValidationKey(_ *jwt.Token) (interface{}, error) {
+func (j *JWT) getValidationKey(token *jwt.Token) (interface{}, error) {
+	if j.jwks != nil {
+		id := token.Header["kid"]
+		algorithm := token.Header["alg"]
+		if id == nil || algorithm == nil {
```
Comment (johakoch) [inline accesscontrol/jwt.go:203]: split these tests?

Comment (johakoch) [reply accesscontrol/jwt.go:203]: If `kid` is optional in JWKS, do we have to require the `kid` header here?

Comment (afflerbach) [reply accesscontrol/jwt.go:203]: At least: `if id == nil || id == "" || …`

File: accesscontrol/jwk.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+			}
+
+			key = certificate.PublicKey.(*rsa.PublicKey)
+		} else if raw.N != nil && raw.E != nil {
+			key = &rsa.PublicKey{
+				N: raw.N.toBigInt(),
+				E: raw.E.toInt(),
+			}
+		} else {
+			fmt.Printf("Ignoring invalid %s key: %q", raw.Kty, raw.Kid)
```
Comment (johakoch) [inline accesscontrol/jwk.go:74]: `fmt.Errorf()`?

Comment (johakoch) [inline accesscontrol/jwk.go:74]: `fmt.Errorf()`?

File: accesscontrol/jwk.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+			return nil
+		}
+
+	default:
+		fmt.Printf("Found unsupported %s key: %q\n", raw.Kty, raw.Kid)
```
Comment (johakoch) [inline accesscontrol/jwk.go:79]: `fmt.Errorf()`?

File: accesscontrol/jwks.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+		ttl:       timetolive,
+	}, nil
+}
+
+func (self *JWKS) GetKeys(kid string) []JWK {
+	var keys []JWK
+
+	if len(self.Keys) == 0 || self.hasExpired() {
+		if err := self.Load(); err != nil {
+			fmt.Printf("Error loading JWKS: %v\n", err)
```
Comment (johakoch) [inline accesscontrol/jwks.go:55]: `fmt.Errorf()`?

File: accesscontrol/jwt.go [hunk 5]
```diff
@@ -167,7 +196,24 @@ func (j *JWT) Validate(req *http.Request) error {
...
+			fmt.Printf("Missing \"kid\" or \"alg\" in JOSE header\n")
```
Comment (johakoch) [inline accesscontrol/jwt.go:204]: `fmt.Errorf()`?

File: accesscontrol/jwt.go [hunk 6]
```diff
@@ -167,7 +196,24 @@ func (j *JWT) Validate(req *http.Request) error {
...
+			return nil, nil
+		}
+		jwk := j.jwks.GetKey(id.(string), algorithm.(string), "sig")
+
+		if jwk == nil {
+			fmt.Printf("No matching %s JWK for kid %q\n", algorithm, id)
```
Comment (johakoch) [inline accesscontrol/jwt.go:210]: `fmt.Errorf()`?

**Review by johakoch — CHANGES_REQUESTED** [review]

File: accesscontrol/jwk.go (hunk 3 above)
Comment (afflerbach) [reply accesscontrol/jwk.go:79]: `Errorf` will be rejected by static analyzer.

File: accesscontrol/jwk.go (hunk 2 above)
Comment (afflerbach) [reply accesscontrol/jwk.go:74]: `Errorf` will be rejected by static analyzer.

**Review by johakoch — DISMISSED** [review]

**Review by johakoch — APPROVED** [review]

**Pull request merged**